internal/indexer/indexer.go

```diff
@@ -443,3 +447,42 @@ func ServiceAccountsByOIDCClaims(obj client.Object) []string {
 	}
 	return refinedClaimValues
 }
+
+// WarehousesByRepoURL is a client.IndexerFunc that indexes Warehouses by the
+// the RepoURLs they are associated with.
+func WarehousesByRepoURL(obj client.Object) []string {
+	warehouse, ok := obj.(*kargoapi.Warehouse)
+	if !ok {
+		return nil
+	}
+
+	var repoURLs []string
+	for _, sub := range warehouse.Spec.Subscriptions {
+		if sub.Git != nil && sub.Git.RepoURL != "" {
+			repoURLs = apppendIfNotExists(repoURLs,
+				git.NormalizeURL(sub.Git.RepoURL),
+			)
+		}
+		if sub.Chart != nil && sub.Chart.RepoURL != "" {
+			repoURLs = apppendIfNotExists(repoURLs,
+				helm.NormalizeChartRepositoryURL(sub.Chart.RepoURL),
+			)
+		}
+		if sub.Image != nil && sub.Image.RepoURL != "" {
+			repoURLs = apppendIfNotExists(repoURLs,
+				// The normalization of Helm chart repository URLs can also be used here
+				// to ensure the uniqueness of the image reference as it does the job of
+				// ensuring lower-casing, etc. without introducing unwanted side effects.
+				helm.NormalizeChartRepositoryURL(sub.Image.RepoURL),
+			)
+		}
+	}
+	return repoURLs
+}
+
+func apppendIfNotExists(ss []string, repoURL string) []string {
+	if !slices.Contains(ss, repoURL) {
+		return append(ss, repoURL)
+	}
+	return ss
+}
```

Comment:
If the same URL were to appear in the list more than once, which is unlikely to begin with, I don't believe it would actually harm anything. I'd probably just kill this bit off.